btree/btree.c

```diff
@@ -0,0 +1,13037 @@
+#define M0_BE_ALLOC_CREDIT_BUF(buf, seg, cred)                               \
+	do { *(buf) = *(buf); (seg) = (seg); *(cred) = *(cred); } while (0)
+
+#define m0_be_ut_tx_init(tx, ut_be)                                          \
+	do { } while (0)
+
+#define m0_be_tx_init(tx,tid,dom,sm_group,persistent,discarded,filler,datum) \
+	do { } while (0)
+
+#define m0_be_tx_prep(tx,credit)                                             \
+	do { } while (0)
+
+#define m0_be_tx_open_sync(tx)  (0)
+
+#define m0_be_tx_open(tx)                                                    \
+	do { } while (0)
+
+#define m0_be_tx_capture(tx,req)                                             \
+	do { } while (0)
+
+#define m0_be_tx_close_sync(tx)                                              \
+	do { } while (0)
+
+#define m0_be_tx_close(tx)                                                   \
+	do { } while (0)
+
+#define m0_be_tx_fini(tx)                                                    \
+	do { } while (0)
+
+#define m0_be_ut_seg_reload(ut_seg)                                          \
+	do { } while (0)
+
+#define m0_be_ut_backend_init(ut_be)                                         \
+	do { } while (0)
+
+#define m0_be_ut_seg_init(ut_seg, ut_be, size)                               \
+	do {                                                                 \
+		M0_ALLOC_PTR(ut_seg->bus_seg);                               \
+		ut_seg->bus_seg->bs_gen = m0_time_now();                     \
+	} while (0)
+
+#define m0_be_ut_seg_fini(ut_seg)                                            \
+	do {                                                                 \
+		m0_free(ut_seg->bus_seg);                                    \
+	} while (0)
+
+#define m0_be_ut_backend_fini(ut_be)                                         \
+	do { } while (0)
+
+#define m0_be_seg_close(ut_seg)                                              \
+	do { } while (0)
+
+#define m0_be_seg_open(ut_seg)                                               \
+	do { } while (0)
+
+#define madvise(rnode, rnode_sz, MADV_NORMAL)                                \
+	-1, errno = ENOMEM
+
+#undef m0_be_engine_tx_size_max
+#define m0_be_engine_tx_size_max(engine, cred, payload_size)                 \
+	do {                                                                 \
+		*(cred) = M0_BE_TX_CREDIT(50, (4 * 1024));            \
+	} while(0)
+
+#endif
+
+/**
+ *  --------------------------------------------
+ *  Section END - BTree Structure and Operations
+ *  --------------------------------------------
+ */
+
+/**
+ *  ---------------------------------------------------
+ *  Section START - BTree Node Structure and Operations
+ *  ---------------------------------------------------
+ */
+
+
+/**
+ * "Address" of a node in a segment.
+ *
+ * Highest 8 bits (56--63) are reserved and must be 0.
+ *
+ * Lowest 4 bits (0--3) contains the node size, see below.
+ *
+ * Next 5 bits (4--8) are reserved and must be 0.
```

Comment:
Any reason for 5 reserved bits? starting fields at boundary of 4 is good from debugging point of view and may be from optimization also.